### PR TITLE
Fp16 training initializers

### PR DESCRIPTION
**Pull request by slayton58**
Adds support for generating and training pfp16 models. Added SGD optimizer for multi-precision trainers and a new callback to data_parallel_model in order to help multi-precision models keep their different copies of parameters in sync during training.

### CODE REVIEW

File: caffe2/python/modeling/initializers.py [hunk 1]
```diff
@@ -29,6 +29,33 @@ def create_param(self, param_name, init_net, shape):
             shape=shape,
         )
 
+class pFP16Initializer(object):
+    def __init__(self, operator_name=None, **kwargs):
+        self.operator_name = operator_name
+        self.operator_kwargs = kwargs
+
+    def update(self, operator_name, kwargs):
```
Comment (salexspb) [inline caffe2/python/modeling/initializers.py]: why not inherit and then reuse base class method?

File: caffe2/python/modeling/initializers.py [hunk 2]
```diff
@@ -29,6 +29,33 @@ def create_param(self, param_name, init_net, shape):
...
+        self.operator_kwargs = kwargs
+
+    def create_param(self, param_name, init_net, shape):
+        # create master fp32 copy
+        param_fp32 = init_net.__getattr__(self.operator_name)(
+            [], param_name+'_fp32', shape=shape,
+            **self.operator_kwargs)
+        # cast to fp16 copy
+        param = init_net.FloatToHalf(
+            param_name+"_fp32", param_name)
```
Comment (salexspb) [inline caffe2/python/modeling/initializers.py]: nit: param_name + "_fp32"

File: caffe2/python/modeling/initializers.py [hunk 3]
```diff
@@ -29,6 +29,33 @@ def create_param(self, param_name, init_net, shape):
...
+        self.operator_name = operator_name
+        self.operator_kwargs = kwargs
+
+    def update(self, operator_name, kwargs):
+        if self.operator_name is not None:
+            raise Exception("Operator name overwrites are not allowed")
+        self.operator_name = operator_name
+        self.operator_kwargs = kwargs
+
+    def create_param(self, param_name, init_net, shape):
```
Comment (salexspb) [inline caffe2/python/modeling/initializers.py:40]: you could call base class method and then add a conversion part + modify ParameterInfo object returned from the base class.

File: caffe2/python/modeling/parameter_info.py [hunk 4]
```diff
@@ -15,7 +15,7 @@ class ParameterType(object):
 class ParameterInfo(object):
     def __init__(
             self, param_id, param, key=None, shape=None, length=None,
-            grad=None):
+            grad=None, blob_copy=None):
```
Comment (salexspb) [inline caffe2/python/modeling/parameter_info.py:18]: could you add a comment about what blob_copy is for? Also how about calling it "float_copy" ?

File: caffe2/python/modeling/initializers.py [hunk 5]
```diff
@@ -29,6 +29,33 @@ def create_param(self, param_name, init_net, shape):
...
+            **self.operator_kwargs)
+        # cast to fp16 copy
+        param = init_net.FloatToHalf(
+            param_name+"_fp32", param_name)
+
+        return ParameterInfo(
+            param_id=None,
+            param=param,
+            shape=shape,
+            blob_copy={'float' : param_fp32}
```
Comment (salexspb) [inline caffe2/python/modeling/initializers.py]: does it have to be a dict? Is there a usecase now where we need it? If not, should we make it simpler for now?

File: caffe2/python/optimizer.py [hunk 6]
```diff
@@ -140,7 +139,7 @@ def _run(self, net, param_init_net, param_info):
 
         if self.momentum > 0:
             momentum_data = param_init_net.ConstantFill(
-                param, str(param) + "_momentum", value=0.)
+                [param], param + '_momentum', value=0.)
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: I think @akyrola 's diff is almost ready which should solve this.

File: caffe2/python/optimizer.py [hunk 7]
```diff
@@ -168,6 +167,61 @@ def scale_learning_rate(self, scale):
...
+        net.MomentumSGDUpdate(
+            [grad+"_fp32", momentum_data, lr, param_fp32],
+            [grad, momentum_data, param_fp32],
+            momentum=self.momentum,
+            nesterov=0)
+
+        # Copy updated param back to fp16
+        net.FloatToHalf(param_fp32, param)
+
+    def scale_learning_rate(self, scale):
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: do you use this?

File: caffe2/python/optimizer.py [hunk 8]
```diff
@@ -168,6 +167,61 @@ def scale_learning_rate(self, scale):
...
+
+        assert not isinstance(grad, core.GradientSlice), \
+                "Doesn't support sparse gradients"
+
+        # Copy gradient to fp32
+        net.HalfToFloat(grad, grad+"_fp32")
+
+        # update (fused) in fp32
+        net.MomentumSGDUpdate(
+            [grad+"_fp32", momentum_data, lr, param_fp32],
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: nit: spaces around +

File: caffe2/python/optimizer.py [hunk 9]
```diff
@@ -168,6 +167,61 @@ def scale_learning_rate(self, scale):
...
+
+        momentum_data = param_init_net.ConstantFill(
+            [param], param + "_momentum", value=0.)
+        self._aux_params.local.append(momentum_data)
+
+        assert not isinstance(grad, core.GradientSlice), \
+                "Doesn't support sparse gradients"
+
+        # Copy gradient to fp32
+        net.HalfToFloat(grad, grad+"_fp32")
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: spaces around +

Comment (salexspb) [inline caffe2/python/optimizer.py]: nit: spaces. Also you could assign this to a BlobReference and use it later instead of relying on string matching

File: caffe2/python/optimizer.py [hunk 10]
```diff
@@ -168,6 +167,61 @@ def scale_learning_rate(self, scale):
...
+        )
+
+        ONE = param_init_net.ConstantFill([], "ONE", shape=[1], value=1.0)
+        self._aux_params.shared.append(ONE)
+
+        momentum_data = param_init_net.ConstantFill(
+            [param], param + "_momentum", value=0.)
+        self._aux_params.local.append(momentum_data)
+
+        assert not isinstance(grad, core.GradientSlice), \
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: can we reuse more code from the SgdOptimizer? And also then get this one for free. At least calls till this line could be split to a function in SgdOptimizer and here you could call it (no need to change other optimizers).

File: caffe2/python/optimizer.py (hunk 7 above)
Comment (salexspb) [inline caffe2/python/optimizer.py]: do you use this anywhere?

File: caffe2/python/data_parallel_model.py [hunk 11]
```diff
@@ -205,6 +211,15 @@ def Parallelize_GPU(
         devices, model_helper_obj, model_helper_obj.param_init_net, sync_names
     )
 
+    # Handle any operations that need to be done after parameter sync
+    # i.e. making sure multi-precision copies of parameters are up-to-date
+    if post_sync_builder_fun is not None:
```
Comment (salexspb) [inline caffe2/python/data_parallel_model.py:233]: I would like @akyrola  to have a look into this part as I am not a DPM expert

**Review by salexspb — CHANGES_REQUESTED** [review]

**Comment by salexspb** [comment]
Btw, @akyrola also has a diff out which makes SgdOptimizer use fused MomentumSGDUpdate which probably should allow you to reuse SgdOptimizer code (his diff is accepted)

File: caffe2/python/modeling/initializers.py (hunk 1 above)
Comment (slayton58) [reply caffe2/python/modeling/initializers.py]: Because I don't like python's inheritance syntax :) Fixed.

File: caffe2/python/modeling/initializers.py (hunk 2 above)
Comment (slayton58) [reply caffe2/python/modeling/initializers.py]: Fixed.

File: caffe2/python/modeling/initializers.py (hunk 3 above)
Comment (slayton58) [reply caffe2/python/modeling/initializers.py:40]: I felt this was cleaner (plus see previous comment about not liking python's inheritance) - it's very explicit about creating the different precisions.

File: caffe2/python/modeling/initializers.py (hunk 5 above)
Comment (slayton58) [reply caffe2/python/modeling/initializers.py]: Doesn't have to be, but it makes sense to me that it is a dict - you also need to know what the different multi-precision types are when updating parameters, so this solves that problem too.

File: caffe2/python/modeling/parameter_info.py (hunk 4 above)
Comment (slayton58) [reply caffe2/python/modeling/parameter_info.py:18]: See above - it's intended as a general place to store copies, not just float.

File: caffe2/python/optimizer.py (hunk 7 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: `net.FloatToHalf`? Yes. You update the fp32 copy of the weights, then truncate them down to fp16 for the next fprop + bprop pass

File: caffe2/python/optimizer.py (hunk 8 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Fixed.

File: caffe2/python/optimizer.py (hunk 7 above)
Comment (salexspb) [reply caffe2/python/optimizer.py]: no, the function scale_learning_rate. I didn't see its usage in the lines you changed.

Comment (salexspb) [reply caffe2/python/optimizer.py]: omg, I hate github commenting capabilities.. Can I select a single line or a set of lines I need? Here I click on the line and it shows up as 4.. And I can't select code blocks.

Comment (slayton58) [reply caffe2/python/optimizer.py]: No, we don't use it anywhere, I assumed it was necessary for the interface, and so had to be provided.

File: caffe2/python/optimizer.py (hunk 10 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: I don't really see how apart from abstracting out some more of the details (which I consider out of the scope of this patch set) -- I already call out to `SgdOptimizer` in the case of fp32 parameters.

File: caffe2/python/optimizer.py (hunk 9 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Fixed.

Comment (slayton58) [reply caffe2/python/optimizer.py]: Done.

File: caffe2/python/modeling/initializers.py [hunk 12]
```diff
@@ -29,6 +29,32 @@ def create_param(self, param_name, init_net, shape):
             shape=shape,
         )
 
+class pFP16Initializer(Initializer):
+    def __init__(self, operator_name=None, **kwargs):
+        super(pFP16Initializer, self).__init__(operator_name, **kwargs)
+
+    def update(self, operator_name, kwargs):
```
Comment (salexspb) [inline caffe2/python/modeling/initializers.py]: why do you still have this method?

File: caffe2/python/modeling/initializers.py (hunk 3 above)
Comment (salexspb) [reply caffe2/python/modeling/initializers.py:40]: What do you mean? An initializer here is essentially a normal one + fp16 conversion, why not reuse more code?

File: caffe2/python/modeling/parameter_info.py (hunk 4 above)
Comment (salexspb) [reply caffe2/python/modeling/parameter_info.py:18]: Thats not clear why is copy needed. Could you add a comment block, please?

File: caffe2/python/optimizer.py (hunk 6 above)
Comment (salexspb) [reply caffe2/python/optimizer.py]: so do you need this change in your diff ? isn't it equivalent to what it was before?

File: caffe2/python/optimizer.py (hunk 7 above)
Comment (salexspb) [reply caffe2/python/optimizer.py]: why? if you inherit you need only to provide things you override

File: caffe2/python/optimizer.py [hunk 13]
```diff
@@ -379,6 +442,13 @@ def build_sgd(model, base_learning_rate, **kwargs):
     sgd_optimizer = SgdOptimizer(base_learning_rate, **kwargs)
     return _build(model, sgd_optimizer)
 
+def build_multi_precision_sgd(model, base_learning_rate, **kwargs):
+    sgd_optimizer = MultiPrecisionSgdOptimizer(base_learning_rate, **kwargs)
+    return _build(model, sgd_optimizer)
+    #for param_info in model.GetOptimizationParamInfo():
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: remove extra comments?

File: caffe2/python/optimizer.py [hunk 14]
```diff
@@ -177,6 +176,70 @@ def scale_learning_rate(self, scale):
...
+            learning_rate_blob=str(param) + "_lr",
+            policy=self.policy,
+            **(self.init_kwargs)
+        )
+
+        dev = scope.CurrentDeviceScope()
+        if dev is None:
+            dev = core.DeviceOption(caffe2_pb2.CPU)
+
+        ONE = param_init_net.ConstantFill(
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: why do you need ONE? I don't see it being used below

File: caffe2/python/optimizer.py [hunk 15]
```diff
@@ -177,6 +176,70 @@ def scale_learning_rate(self, scale):
...
+            return SgdOptimizer._run(self, net, param_init_net, param_info)
+
+        grad = param_info.grad
+        if self.base_learning_rate <= 0:
+            return
+
+        lr, _ = self.build_lr(
+            net, param_init_net,
+            base_learning_rate=-self.base_learning_rate,
+            learning_rate_blob=str(param) + "_lr",
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: why do you need str() on param here?  when you sum up an object and a string it would call __str__ on it automatically

**Review by salexspb — CHANGES_REQUESTED** [review]

File: caffe2/python/modeling/initializers.py (hunk 3 above)
Comment (slayton58) [reply caffe2/python/modeling/initializers.py:40]: Because right now it'll look like:

```
param_fp32_info = super(Initializer, self).create_param(
        param_name + "fp32",
        init_net,
        shape)

param = init_net.FloatToHalf(param_fp32_info.blob,
                                                param_name)

return ParameterInfo(
        param_id=None,
        param=param,
        shape=shape,
        blob_copy={'float' : param_fp32_info.blob})
```

That's not really cleaner in my opinion

File: caffe2/python/modeling/parameter_info.py (hunk 4 above)
Comment (slayton58) [reply caffe2/python/modeling/parameter_info.py:18]: I already added that in the body of the method. Would you prefer it prefixing the function? Inline with the signature?

File: caffe2/python/optimizer.py (hunk 6 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: I believe this is actually still needed, otherwise I think the namescope will still be applied.

File: caffe2/python/optimizer.py (hunk 13 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Done.

File: caffe2/python/optimizer.py (hunk 14 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Removed.

File: caffe2/python/data_parallel_model.py (hunk 11 above)
Comment (akyrola) [reply caffe2/python/data_parallel_model.py:233]: This is fine

File: caffe2/python/optimizer.py (hunk 6 above)
Comment (akyrola) [reply caffe2/python/optimizer.py]: Actually not needed anymore, if you use optimizer_builder_fun with data_parallel_model, as that is called without any Namescopes.

File: caffe2/python/optimizer.py [hunk 16]
```diff
@@ -185,6 +184,53 @@ def scale_learning_rate(self, scale):
...
+
+        # Copy gradient to fp32
+        grad_fp32 = net.HalfToFloat(grad, grad + "_fp32")
+
+        # update (fused) in fp32
+        net.MomentumSGDUpdate(
+            [grad_fp32, momentum_data, lr, param_fp32],
+            [grad, momentum_data, param_fp32],
+            momentum=self.momentum,
+            nesterov=0)
```
Comment (akyrola) [inline caffe2/python/optimizer.py]: nesterov=self.nesterov

File: caffe2/python/optimizer.py [hunk 17]
```diff
@@ -185,6 +184,53 @@ def scale_learning_rate(self, scale):
         self.base_learning_rate *= scale
         return
 
+class MultiPrecisionSgdOptimizer(SgdOptimizer):
+    def __init__(self, base_learning_rate, momentum, policy, **kwargs):
+        super(SgdOptimizer, self).__init__() # self, base_learning_rate, **kwargs)
+        self.base_learning_rate = base_learning_rate
+        self.momentum = momentum
+        self.policy = policy
+        self.init_kwargs = kwargs
```
Comment (akyrola) [inline caffe2/python/optimizer.py]: self.nesterov = nesterov

and add nesterov as parameter, default=0

File: caffe2/python/optimizer.py [hunk 18]
```diff
@@ -387,6 +433,10 @@ def build_sgd(model, base_learning_rate, **kwargs):
     sgd_optimizer = SgdOptimizer(base_learning_rate, **kwargs)
     return _build(model, sgd_optimizer)
 
+def build_multi_precision_sgd(model, base_learning_rate, **kwargs):
+    sgd_optimizer = MultiPrecisionSgdOptimizer(base_learning_rate, **kwargs)
```
Comment (akyrola) [inline caffe2/python/optimizer.py]: nit:
  multiprec_sgd_optimizer = ...

File: caffe2/python/optimizer.py (hunk 16 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Fixed.

File: caffe2/python/optimizer.py (hunk 18 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Fixed.

File: caffe2/python/data_parallel_model.py (hunk 11 above)
Comment (akyrola) [reply caffe2/python/data_parallel_model.py:233]: @slayton58  looks like you don't use the post_sync_builder_fun anymore? Perhaps because optimizer_builder_fun is now possible? 

In that case, let's remove it.

File: caffe2/python/helpers/conv.py [hunk 19]
```diff
@@ -18,6 +18,8 @@ def _ConvBase(
     kernel,
     weight_init=None,
     bias_init=None,
+    WeightInitializer=None,
+    BiasInitializer=None,
```
Comment (akyrola) [inline caffe2/python/helpers/conv.py:22]: why different casing for WeightInitializer and BiasInitializer?

File: caffe2/python/helpers/conv.py [hunk 20]
```diff
@@ -54,20 +54,23 @@ def _ConvBase(
...
+    BiasInitializer = initializers.update_initializer(
+        BiasInitializer, bias_init, ("ConstantFill", {})
+    )
+
     if model.init_params:
-        weight = model.param_init_net.__getattr__(weight_init[0])(
-            [],
-            blob_out + '_w',
+        weight = model.create_param(
+            param_name=blob_out+'_w',
```
Comment (akyrola) [inline caffe2/python/helpers/conv.py]: nit: spaces around +

File: caffe2/python/helpers/conv.py [hunk 21]
```diff
@@ -54,20 +54,23 @@ def _ConvBase(
...
             shape=weight_shape,
-            **weight_init[1]
-        )
+            initializer=WeightInitializer)
         if use_bias:
-            bias = model.param_init_net.__getattr__(bias_init[0])(
-                [],
-                blob_out + '_b',
+            bias = model.create_param(
+                param_name=blob_out+'_b',
```
Comment (akyrola) [inline caffe2/python/helpers/conv.py]: ditto

File: caffe2/python/optimizer.py [hunk 22]
```diff
@@ -157,7 +156,7 @@ def _run(self, net, param_init_net, param_info):
 
         if self.momentum > 0:
             momentum_data = param_init_net.ConstantFill(
-                param, str(param) + "_momentum", value=0.)
+                param, str(param) + '_momentum', value=0.)
```
Comment (akyrola) [inline caffe2/python/optimizer.py]: unnecessary change?

File: caffe2/python/optimizer.py [hunk 23]
```diff
@@ -387,6 +434,12 @@ def build_sgd(model, base_learning_rate, **kwargs):
     sgd_optimizer = SgdOptimizer(base_learning_rate, **kwargs)
     return _build(model, sgd_optimizer)
 
+def build_multi_precision_sgd(model, base_learning_rate, **kwargs):
+    multi_prec_sgd_optimizer = MultiPrecisionSgdOptimizer(
+            base_learning_rate, **kwargs
+    )
+    return _build(model, sgd_optimizer)
```
Comment (akyrola) [inline caffe2/python/optimizer.py]: doesn't compile... sgd_optimizer not defined anymore

**Review by akyrola — COMMENTED** [review]
A couple of more comments, and I don't think this currently works (see comment).

File: caffe2/python/data_parallel_model.py (hunk 11 above)
Comment (slayton58) [reply caffe2/python/data_parallel_model.py:233]: No, I still need it. `optimizer_builder_fun` happens here: https://github.com/caffe2/caffe2/blob/master/caffe2/python/data_parallel_model.py#L183-L191 while `SyncParams` happens later, here: https://github.com/caffe2/caffe2/blob/master/caffe2/python/data_parallel_model.py#L222-L224

My sync needs to happen after the `SyncParams` call

File: caffe2/python/helpers/conv.py (hunk 19 above)
Comment (slayton58) [reply caffe2/python/helpers/conv.py:22]: Ask @salexspb :) He wanted the CamelCase for the *Initializer to indicate they're classes being passed in, not objects.

File: caffe2/python/helpers/conv.py (hunk 20 above)
Comment (slayton58) [reply caffe2/python/helpers/conv.py]: Fixed.

File: caffe2/python/helpers/conv.py (hunk 21 above)
Comment (slayton58) [reply caffe2/python/helpers/conv.py]: Fixed.

File: caffe2/python/optimizer.py (hunk 23 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Fixed.

File: caffe2/python/modeling/parameter_info.py [hunk 24]
```diff
@@ -26,6 +26,9 @@ def __init__(
         self.length = max(1, length if length is not None else 1)
         self.grad = grad
         self._cloned_init_net = None
+        # Optionally store equivalent copies of the blob
+        # in different precisions (i.e. half and float copies)
+        self.blob_copy = blob_copy
```
Comment (salexspb) [inline caffe2/python/modeling/parameter_info.py]: could you notice here that it is a dict?

File: caffe2/python/optimizer.py [hunk 25]
```diff
@@ -185,6 +184,54 @@ def scale_learning_rate(self, scale):
...
+        super(SgdOptimizer, self).__init__()
+        self.base_learning_rate = base_learning_rate
+        self.momentum = momentum
+        self.policy = policy
+        self.nesterov = nesterov
+        self.init_kwargs = kwargs
+
+    def _run(self, net, param_init_net, param_info):
+        param = param_info.blob
+        param_fp32 = param_info.blob_copy['float'] if param_info.blob_copy is not None else None
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: could you replace 'float' with DataType enum in core.py?

File: caffe2/python/optimizer.py [hunk 26]
```diff
@@ -185,6 +184,54 @@ def scale_learning_rate(self, scale):
...
+
+        # If we have a straight fp32 parameter, run the base class
+        if param_fp32 == None:
+            return SgdOptimizer._run(self, net, param_init_net, param_info)
+
+        grad = param_info.grad
+        if self.base_learning_rate <= 0:
```
Comment (salexspb) [inline caffe2/python/optimizer.py]: what is this if statement for? IMO this is error prone. is this an input error check? Then should we through otherwise?

File: caffe2/python/optimizer.py [hunk 27]
```diff
@@ -185,6 +184,54 @@ def scale_learning_rate(self, scale):
...
+        self._aux_params.local.append(momentum_data)
+
+        assert not isinstance(grad, core.GradientSlice), \
+                "Doesn't support sparse gradients"
+
+        # Copy gradient to fp32
+        grad_fp32 = net.HalfToFloat(grad, grad + "_fp32")
+
+        # update (fused) in fp32
+        net.MomentumSGDUpdate(
```
Comment (salexspb) [inline caffe2/python/optimizer.py:230]: Not suggesting to do it in this PR, but going forward it would be nice to have a way to convert an optimizer to fp16 format without replicating its  code.  What do you think?

**Review by salexspb — COMMENTED** [review]
Just a few small things and should be good to go!

File: caffe2/python/helpers/conv.py [hunk 28]
```diff
@@ -54,20 +54,23 @@ def _ConvBase(
         weight_shape.extend(kernels)
         weight_shape.append(int(dim_in / group))
 
+    WeightInitializer = initializers.update_initializer(
```
[the remark this answers is not in the document]
Comment (slayton58) [reply caffe2/python/helpers/conv.py]: Done.

File: caffe2/python/modeling/parameter_info.py (hunk 24 above)
Comment (slayton58) [reply caffe2/python/modeling/parameter_info.py]: Done.

File: caffe2/python/optimizer.py (hunk 26 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: A net will generally have some parameters in fp16, some in float (i.e. Resnets where all filters are fp16, but batch norm parameters are all fp32) - this just drops back to the standard fp32 update path if an fp32 parameter is encountered.

File: caffe2/python/optimizer.py (hunk 27 above)
Comment (slayton58) [reply caffe2/python/optimizer.py:230]: It would be - I'd like to discuss that with you in fact.

There's a couple of ways of doing it.

File: caffe2/python/optimizer.py (hunk 26 above)
Comment (salexspb) [reply caffe2/python/optimizer.py]: I am talking here about line 205. It is the one I clicked on when leaving this comment. How do I make commenting here less confusing for you? :)

File: caffe2/python/optimizer.py (hunk 25 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: Done.

File: caffe2/python/optimizer.py (hunk 26 above)
Comment (slayton58) [reply caffe2/python/optimizer.py]: This is to match the behaviour of the SgdOptimizer (line 127)

Comment (salexspb) [reply caffe2/python/optimizer.py]: It has == 0, not <= 0. < 0 sounds like a user error.

Comment (slayton58) [reply caffe2/python/optimizer.py]: I see, typo. Fixed.

**Comment by salexspb** [comment]
Looks good code wise! 

What about testing? We already have a test for initializers and optimizers, could you look into adding these new guys in there?

**Comment by slayton58** [comment]
Added (fairly basic) tests

File: caffe2/python/optimizer_test.py [hunk 29]
```diff
@@ -21,6 +23,19 @@ def check_optimizer(self, optimizer):
             np.testing.assert_allclose(np.array([1.0]), tensor, atol=1e-5)
 
 
+class TestMultiPrecisionSgd(OptimizerTestBase, TestCase):
+    def build_optimizer(self, model):
+        self._skip_gpu = False
+        return build_multi_precision_sgd(model, base_learning_rate=0.1)
```
Comment (salexspb) [inline caffe2/python/optimizer_test.py:31]: will this actually exercise the logic? I think if you don't use an fp16 initializer , then you will just fallback to the normal sgd...

File: caffe2/python/optimizer_test.py [hunk 30]
```diff
@@ -21,6 +23,19 @@ def check_optimizer(self, optimizer):
...
+
+    def check_optimizer(self, optimizer):
```
Comment (salexspb) [inline caffe2/python/optimizer_test.py:33]: could you actually test that we at least have those params with a copy and conversion ops in place?

**Review by salexspb — CHANGES_REQUESTED** [review]
Could you improve the optimizer test to cover the new logic, please?

File: caffe2/python/helpers/conv.py [hunk 31]
```diff
@@ -54,20 +54,23 @@ def _ConvBase(
...
+    )
+    bias_initializer = initializers.update_initializer(
+        BiasInitializer, bias_init, ("ConstantFill", {})
+    )
+
     if model.init_params:
-        weight = model.param_init_net.__getattr__(weight_init[0])(
-            [],
-            blob_out + '_w',
+        weight = model.create_param(
```
Comment (akyrola) [inline caffe2/python/helpers/conv.py:65]: This unfortunately breaks things, as now weight & bias are added twice to params.

Test python/data_parallel_model_test.py -- test_checkpoint_params will fail.

I am uncommiting this for now.

Comment (akyrola) [reply caffe2/python/helpers/conv.py:65]: Sorry, this was mistake by me -- the offending diff came later. Trying to reimport.
